File: 0695-max-area-of-island/0695-max-area-of-island.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
const maxAreaOfIsland = grid => {
    let maxArea = 0;
    const m = grid.length;
    const n = grid[0].length;
    const sizeOfIsland = (i, j) => {
        if (i < 0 || j < 0 || i >= m || j >= n) return 0;
        if (grid[i][j] === 0) return 0;
        grid[i][j] = 0;
        return sizeOfIsland(i + 1, j) + sizeOfIsland(i - 1, j) + sizeOfIsland(i, j + 1) + sizeOfIsland(i, j - 1) + 1;
    }
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            const size = sizeOfIsland(i, j);
            maxArea = Math.max(maxArea, size);
        }
    }
    return maxArea;
};

File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        maxSize = 0
        def sizeIsland(grid, i, j):
            if i < 0 or j < 0 or i >= m or j >= n:
                return 0
            if grid[i][j] == 0:
                return 0
            grid[i][j] = 0
            return sizeIsland(grid, i + 1, j) + sizeIsland(grid, i, j + 1) + sizeIsland(grid, i - 1, j) + sizeIsland(grid, i, j - 1) + 1
            
            
        for i in range(m):
            for j in range(n):
                size = sizeIsland(grid, i, j)
                maxSize = max(maxSize, size)
        return maxSize